bpf/tests/ipsec_from_overlay_test.c

```diff
@@ -0,0 +1,562 @@
+			    __maybe_unused __u32 size,
+			    __maybe_unused __u32 flags)
+{
+	to->remote_ipv4 = v4_node_one;
+	/* 0xfffff is the default SECLABEL */
+	to->tunnel_id = 0xfffff;
+	return 0;
+}
+
+__section("mock-handle-policy")
+int mock_handle_policy(struct __ctx_buff *ctx __maybe_unused)
+{
+	return TC_ACT_REDIRECT;
+}
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PROG_ARRAY);
+	__uint(key_size, sizeof(__u32));
+	__uint(max_entries, 256);
+	__array(values, int());
+} mock_policy_call_map __section(".maps") = {
+	.values = {
+		[DEST_LXC_ID] = &mock_handle_policy,
+	},
+};
+
+#define tail_call_dynamic mock_tail_call_dynamic
+static __always_inline __maybe_unused void
+mock_tail_call_dynamic(struct __ctx_buff *ctx __maybe_unused,
+		       const void *map __maybe_unused, __u32 slot __maybe_unused)
+{
+	tail_call(ctx, &mock_policy_call_map, slot);
+}
+
+static volatile const __u8 *DEST_EP_MAC = mac_three;
+static volatile const __u8 *DEST_NODE_MAC = mac_four;
+
+#include "bpf_overlay.c"
+
+#define FROM_OVERLAY 0
+#define ESP_SEQUENCE 69865
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PROG_ARRAY);
+	__uint(key_size, sizeof(__u32));
+	__uint(max_entries, 1);
+	__array(values, int());
+} entry_call_map __section(".maps") = {
+	.values = {
+		[FROM_OVERLAY] = &cil_from_overlay,
+	},
+};
+
+PKTGEN("tc", "ipv4_with_encrypt_ipsec_from_overlay")
+int ipv4_with_encrypt_ipsec_from_overlay_pktgen(struct __ctx_buff *ctx)
+{
+	struct pktgen builder;
+	struct ethhdr *l2;
+	struct iphdr *l3;
+	struct ip_esp_hdr *l4;
+	void *data;
+
+	pktgen__init(&builder, ctx);
+
+	l2 = pktgen__push_ethhdr(&builder);
+	if (!l2)
+		return TEST_ERROR;
+	ethhdr__set_macs(l2, (__u8 *)mac_one, (__u8 *)mac_two);
+
+	l3 = pktgen__push_default_iphdr(&builder);
+	if (!l3)
+		return TEST_ERROR;
+	l3->saddr = v4_pod_one;
+	l3->daddr = v4_pod_two;
+
+	l4 = pktgen__push_esphdr(&builder);
+	if (!l4)
+		return TEST_ERROR;
+	if ((void *)l4 + sizeof(struct ip_esp_hdr) > ctx_data_end(builder.ctx))
+		return 0;
+	l4->spi = ENCRYPT_KEY;
+	l4->seq_no = ESP_SEQUENCE;
+
+	data = pktgen__push_data(&builder, default_data, sizeof(default_data));
+	if (!data)
+		return TEST_ERROR;
+
+	pktgen__finish(&builder);
+	return 0;
+}
+
+SETUP("tc", "ipv4_with_encrypt_ipsec_from_overlay")
+int ipv4_with_encrypt_ipsec_from_overlay_setup(struct __ctx_buff *ctx)
+{
+	tail_call_static(ctx, &entry_call_map, FROM_OVERLAY);
+	return TEST_ERROR;
+}
+
+CHECK("tc", "ipv4_with_encrypt_ipsec_from_overlay")
+int ipv4_with_encrypt_ipsec_from_overlay_check(__maybe_unused const struct __ctx_buff *ctx)
+{
+	void *data;
+	void *data_end;
+	__u32 *status_code;
+	struct ethhdr *l2;
+	struct iphdr *l3;
+	struct ip_esp_hdr *l4;
+	__u8 *payload;
+
+	test_init();
+
+	data = (void *)(long)ctx->data;
+	data_end = (void *)(long)ctx->data_end;
+
+	if (data + sizeof(*status_code) > data_end)
+		test_fatal("status code out of bounds");
+
+	status_code = data;
+	assert(*status_code == CTX_ACT_OK);
+	assert(ctx->mark == 0xd00);
```

Comment:
can we also test `skb->pkt_type` here? (potentially other places too?)

Reply:
Julian, your comment made me find this very interesting issue: the context bytes passed to BPF_PROG_TEST_RUN aren't used as skb metadata unconditionally; before I said "briefly we can expect `memcpy(skb, ctx)` happened in kernel, but that's not true.

I did some probe and found only a few fields are copied, briefly speaking it's implemented like:

```c
struct sk_buff *skb, struct __sk_buff *__skb;
memcpy(__skb, ctx);
skb->mark = __skb->mark;
skb->priority = __skb->priority;
skb->skb_iif = __skb->ingress_ifindex;
skb->tstamp = __skb->tstamp;
memcpy(&cb->data, __skb->cb, QDISC_CB_PRIV_LEN);
...
```

skb->pkt_type won't be copied, so I can't check the skb->pkt_type in our bpt unittest.

The source code for your reference: https://github.com/torvalds/linux/blob/5f63595ebd82f56a2dd36ca013dd7f5ff2e2416a/net/bpf/test_run.c#L1079-L1083, that `convert___skb_to_skb` function is called by `bpf_prog_test_run_skb`.

Reply:
For this PR, I'll mock the `ctx_change_type` to make sure the test cases run as expected. Thank you!

Reply:
This `convert_skb_to___skb` is even more clear to tell what fields are copied: https://github.com/torvalds/linux/blob/5f63595ebd82f56a2dd36ca013dd7f5ff2e2416a/net/bpf/test_run.c#L1103

```
	struct qdisc_skb_cb *cb = (struct qdisc_skb_cb *)skb->cb;

	if (!__skb)
		return;

	__skb->mark = skb->mark;
	__skb->priority = skb->priority;
	__skb->ingress_ifindex = skb->skb_iif;
	__skb->ifindex = skb->dev->ifindex;
	__skb->tstamp = skb->tstamp;
	memcpy(__skb->cb, &cb->data, QDISC_CB_PRIV_LEN);
	__skb->wire_len = cb->pkt_len;
	__skb->gso_segs = skb_shinfo(skb)->gso_segs;
	__skb->hwtstamp = skb_shinfo(skb)->hwtstamps.hwtstamp;
```

This is also called inside `bpf_prog_test_run_skb`: https://github.com/torvalds/linux/blob/5f63595ebd82f56a2dd36ca013dd7f5ff2e2416a/net/bpf/test_run.c#L1243

Reply:
![image](https://github.com/cilium/cilium/assets/6435624/b8622dbe-5e31-4da7-97fc-24d01a87f45b)
According to the blame info and the commit which introduced these conversion functions, I don't see reasons to forbid us copying skb->pkt_gen in bpf_prog_test_run. I'll see what I can do in future, but now let it be.

Reply:
Thanks for digging, that's perfectly fine then :+1: